File: src-tauri/src/csv_export.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;
use tauri::AppHandle;

#[derive(Debug, Serialize, Deserialize)]
pub struct ExportResult {
    pub success: bool,
    pub file_path: String,
    pub message: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ExportSettings {
    pub export_directory: String,
    pub auto_open: bool,
    pub include_timestamp: bool,
}

/// Export CSV data to the configured directory
#[tauri::command]
pub async fn export_csv_to_directory(
    app: AppHandle,
    filename: String,
    csv_content: String,
) -> Result<ExportResult, String> {
    // Get export settings from database
    let settings = match crate::settings_db::get_export_settings(&app).await {
        Ok(s) => s,
        Err(e) => {
            return Err(format!("Failed to get export settings: {}", e));
        }
    };

    // Build the full file path
    let export_dir = PathBuf::from(&settings.export_directory);

    // Create directory if it doesn't exist
    if let Err(e) = fs::create_dir_all(&export_dir) {
        return Err(format!("Failed to create export directory: {}", e));
    }

    // Add timestamp to filename if configured
    let final_filename = if settings.include_timestamp {
        let timestamp = chrono::Local::now().format("%Y%m%d_%H%M%S");
        let name_parts: Vec<&str> = filename.rsplitn(2, '.').collect();
        if name_parts.len() == 2 {
            format!("{}_{}.{}", name_parts[1], timestamp, name_parts[0])
        } else {
            format!("{}_{}", filename, timestamp)
        }
    } else {
        filename.clone()
    };

    let file_path = export_dir.join(&final_filename);

    // Write CSV content to file
    match fs::write(&file_path, csv_content.as_bytes()) {
        Ok(_) => {
            let path_str = file_path.to_string_lossy().to_string();

            // Auto-open if configured
            if settings.auto_open {
                #[cfg(target_os = "windows")]
                {
                    let _ = std::process::Command::new("cmd")
                        .args(&["/C", "start", "", &path_str])
                        .spawn();
                }

                #[cfg(target_os = "macos")]
                {
                    let _ = std::process::Command::new("open").arg(&path_str).spawn();
                }

                #[cfg(target_os = "linux")]
                {
                    let _ = std::process::Command::new("xdg-open")
                        .arg(&path_str)
                        .spawn();
                }
            }

            Ok(ExportResult {
                success: true,
                file_path: path_str,
                message: format!("CSV exported successfully to {}", final_filename),
            })
        }
        Err(e) => Err(format!("Failed to write CSV file: {}", e)),
    }
}

/// Export multiple CSV files in a batch
#[tauri::command]
pub async fn export_csv_batch(
    app: AppHandle,
    exports: Vec<(String, String)>, // Vec of (filename, content)
) -> Result<Vec<ExportResult>, String> {
    let mut results = Vec::new();

    for (filename, content) in exports {
        match export_csv_to_directory(app.clone(), filename.clone(), content).await {
            Ok(result) => results.push(result),
            Err(e) => results.push(ExportResult {
                success: false,
                file_path: String::new(),
                message: format!("Failed to export {}: {}", filename, e),
            }),
        }
    }

    Ok(results)
}

/// Get the current export directory
#[tauri::command]
pub async fn get_export_directory(app: AppHandle) -> Result<String, String> {
    match crate::settings_db::get_export_settings(&app).await {
        Ok(settings) => Ok(settings.export_directory),
        Err(e) => Err(format!("Failed to get export directory: {}", e)),
    }
}

/// Open the export directory in file explorer
#[tauri::command]
pub async fn open_export_directory(app: AppHandle) -> Result<(), String> {
    let settings = crate::settings_db::get_export_settings(&app)
        .await
        .map_err(|e| format!("Failed to get export settings: {}", e))?;

    let export_dir = PathBuf::from(&settings.export_directory);

    // Create directory if it doesn't exist
    fs::create_dir_all(&export_dir)
        .map_err(|e| format!("Failed to create export directory: {}", e))?;

    let path_str = export_dir.to_string_lossy().to_string();

    #[cfg(target_os = "windows")]
    {
        std::process::Command::new("explorer")
            .arg(&path_str)
            .spawn()
            .map_err(|e| format!("Failed to open directory: {}", e))?;
    }

    #[cfg(target_os = "macos")]
    {
        std::process::Command::new("open")
            .arg(&path_str)
            .spawn()
            .map_err(|e| format!("Failed to open directory: {}", e))?;
    }

    #[cfg(target_os = "linux")]
    {
        std::process::Command::new("xdg-open")
            .arg(&path_str)
            .spawn()
            .map_err(|e| format!("Failed to open directory: {}", e))?;
    }

    Ok(())
}

File: lib/tauri/csv-export.ts
/**
 * Tauri CSV Export Utilities
 *
 * TypeScript bindings for Rust-powered CSV export functionality
 * with SQLite-backed settings storage
 */

import { invoke } from "@tauri-apps/api/core";

export interface ExportResult {
  success: boolean;
  file_path: string;
  message: string;
}

export interface ExportSettings {
  export_directory: string;
  auto_open: boolean;
  include_timestamp: boolean;
}

/**
 * Export CSV data to the configured directory
 *
 * @param filename - Name of the CSV file (e.g., "students.csv")
 * @param csvContent - CSV content as a string
 * @returns ExportResult with success status and file path
 *
 * @example
 * ```ts
 * const result = await exportCSVToDirectory("students.csv", csvData);
 * if (result.success) {
 *   console.log(`Exported to: ${result.file_path}`);
 * }
 * ```
 */
export async function exportCSVToDirectory(
  filename: string,
  csvContent: string,
): Promise<ExportResult> {
  try {
    return await invoke<ExportResult>("export_csv_to_directory", {
      filename,
      csvContent,
    });
  } catch (error) {
    return {
      success: false,
      file_path: "",
      message: `Export failed: ${error}`,
    };
  }
}

/**
 * Export multiple CSV files in a batch operation
 *
 * @param exports - Array of [filename, content] tuples
 * @returns Array of ExportResults for each file
 *
 * @example
 * ```ts
 * const results = await exportCSVBatch([
 *   ["students.csv", studentData],
 *   ["teachers.csv", teacherData],
 * ]);
 * ```
 */
export async function exportCSVBatch(
  exports: [string, string][],
): Promise<ExportResult[]> {
  try {
    return await invoke<ExportResult[]>("export_csv_batch", {
      exports,
    });
  } catch (error) {
    console.error("Batch export failed:", error);
    return exports.map(() => ({
      success: false,
      file_path: "",
      message: `Export failed: ${error}`,
    }));
  }
}

/**
 * Get the current export directory path
 *
 * @returns The configured export directory path
 */
export async function getExportDirectory(): Promise<string> {
  try {
    return await invoke<string>("get_export_directory");
  } catch (error) {
    console.error("Failed to get export directory:", error);
    return "./exports";
  }
}

/**
 * Open the export directory in the system file explorer
 */
export async function openExportDirectory(): Promise<void> {
  try {
    await invoke("open_export_directory");
  } catch (error) {
    console.error("Failed to open export directory:", error);
    throw error;
  }
}

/**
 * Get current export settings from SQLite database
 *
 * @returns Current export settings
 */
export async function getExportSettings(): Promise<ExportSettings> {
  try {
    return await invoke<ExportSettings>("get_settings");
  } catch (error) {
    console.error("Failed to get settings:", error);
    // Return defaults
    return {
      export_directory: "./exports",
      auto_open: false,
      include_timestamp: true,
    };
  }
}

/**
 * Update export settings in SQLite database
 *
 * @param settings - New settings to save
 *
 * @example
 * ```ts
 * await updateExportSettings({
 *   export_directory: "C:/MyExports",
 *   auto_open: true,
 *   include_timestamp: false,
 * });
 * ```
 */
export async function updateExportSettings(
  settings: ExportSettings,
): Promise<void> {
  try {
    await invoke("update_settings", { settings });
  } catch (error) {
    console.error("Failed to update settings:", error);
    throw error;
  }
}

/**
 * Reset export settings to defaults
 *
 * @returns Default settings
 */
export async function resetExportSettings(): Promise<ExportSettings> {
  try {
    return await invoke<ExportSettings>("reset_settings");
  } catch (error) {
    console.error("Failed to reset settings:", error);
    throw error;
  }
}

/**
 * Open a folder picker dialog to select export directory
 * Updates settings automatically if a folder is selected
 *
 * @returns Selected directory path or null if cancelled
 *
 * @example
 * ```ts
 * const newDir = await selectExportDirectory();
 * if (newDir) {
 *   console.log(`Export directory changed to: ${newDir}`);
 * }
 * ```
 */
export async function selectExportDirectory(): Promise<string | null> {
  try {
    return await invoke<string | null>("select_export_directory");
  } catch (error) {
    console.error("Failed to select directory:", error);
    throw error;
  }
}

/**
 * Check if running in Tauri environment
 * CSV export features only work in Tauri app
 */
export function isExportAvailable(): boolean {
  return typeof window !== "undefined" && "__TAURI__" in window;
}

/**
 * Helper to convert data array to CSV string
 *
 * @param data - Array of objects to convert
 * @param headers - Optional custom headers (defaults to object keys)
 * @returns CSV string
 *
 * @example
 * ```ts
 * const data = [
 *   { name: "John", age: 30 },
 *   { name: "Jane", age: 25 },
 * ];
 * const csv = arrayToCSV(data);
 * await exportCSVToDirectory("data.csv", csv);
 * ```
 */
export function arrayToCSV<T extends Record<string, unknown>>(
  data: T[],
  headers?: string[],
): string {
  if (data.length === 0) return "";

  const cols = headers || Object.keys(data[0]);
  const headerRow = cols.map(escapeCSVField).join(",");

  const rows = data.map((row) =>
    cols
      .map((col) => {
        const value = row[col];
        return escapeCSVField(String(value ?? ""));
      })
      .join(","),
  );

  return [headerRow, ...rows].join("\n");
}

/**
 * Escape CSV field for proper formatting
 */
function escapeCSVField(field: string): string {
  if (field.includes(",") || field.includes('"') || field.includes("\n")) {
    return `"${field.replace(/"/g, '""')}"`;
  }
  return field;
}
